test/integration/ctl/ctl_create_cr_test.go

```diff
@@ -35,7 +36,239 @@ import (
+// but before the CR is successfully created, then we still want to check the validity of the private key.
+
+// TestCtlCreateCRBehaviourBeforeAnythingIsCreated tests the behaviour in the case where the command fails
+// before anything is created, i.e. before private key is successfully written to file.
+func TestCtlCreateCRBehaviourBeforeAnythingIsCreated(t *testing.T) {
+	config, stopFn := framework.RunControlPlane(t)
+	defer stopFn()
+
+	// Build clients
+	_, _, cmCl, _ := framework.NewClients(t, config)
+
+	testdataPath := getTestDataDir(t)
+
+	tests := map[string]Test{
+		"conflicting namespaces defined in flag and file": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr3Name},
+			inputNamespace: ns2,
+			keyFilename:    "",
+			expValidateErr: false,
+			expRunErr:      true,
+			expErrMsg:      "error when getting infos out of Result: the namespace from the provided object \"testns-1\" does not match the namespace \"testns-2\". You must pass '--namespace=testns-1' to perform this operation.",
+		},
+		"file passed in defines resource other than certificate": {
+			inputFile:      testdataPath + "create_cr_issuer.yaml",
+			inputArgs:      []string{cr4Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expValidateErr: false,
+			expRunErr:      true,
+			expErrMsg:      "decoded object is not a v1alpha2 Certificate",
+		},
+		"CR name not passed as arg": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expValidateErr: true,
+			expRunErr:      false,
+			expErrMsg:      "the name of the CertificateRequest to be created has to be provided as argument",
+		},
+	}
+
+	for name, test := range tests {
+		// Run ctl create cr command with input options
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			cleanUpFunc := setupPathForTest(t)
+			defer cleanUpFunc()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
+				CMClient:         cmCl,
+				RESTConfig:       config,
+				IOStreams:        streams,
+				CmdNamespace:     test.inputNamespace,
+				EnforceNamespace: test.inputNamespace != "",
+				InputFilename:    test.inputFile,
+				KeyFilename:      test.keyFilename,
+				CertFileName:     test.certFilename,
+			}
+
+			// Validating args and flags
+			err := opts.Validate(test.inputArgs)
+			if err != nil {
+				if !test.expValidateErr {
+					t.Errorf("got unexpected error when validating args and flags: %v", err)
+				} else if err.Error() != test.expErrMsg {
+					t.Errorf("got unexpected error when validating args and flags, expected: %v; actual: %v", test.expErrMsg, err)
+				}
+				return
+			} else {
+				// got no error
+				if test.expValidateErr {
+					t.Errorf("expected but got no error validating args and flags")
+				}
+			}
```

Comment:
I would add a TODO item suggesting we follow this up by adding unit tests for the `opts.Validate` in the package where it is defined.
And generally, I'd try and do more unit tests for error conditions and leave integration tests for mostly success cases.